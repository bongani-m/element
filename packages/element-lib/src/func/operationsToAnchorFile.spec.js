const element = require('../../index');

const {
  storage,
  primaryKeypair,
  primaryKeypair2,
  secondaryKeypair,
  recoveryKeypair,
  recoveryKeypair2,
} = require('../__tests__/__fixtures__');

// We need only support 1 op per did.
// NOTE: These tests, do not test syncing, so validation does not occurs for order or signatures.
// See: https://github.com/decentralized-identity/sidetree/issues/173
describe('operationsToAnchorFile', () => {
  it('create', async () => {
    const encodedPayload = element.func.encodeJson({
      '@context': 'https://w3id.org/did/v1',
      publicKey: [
        {
          id: '#primary',
          type: 'Secp256k1VerificationKey2018',
          publicKeyHex: primaryKeypair.publicKey,
        },
        {
          id: '#recovery',
          type: 'Secp256k1VerificationKey2018',
          publicKeyHex: recoveryKeypair.publicKey,
        },
      ],
    });
    const signature = element.func.signEncodedPayload(encodedPayload, primaryKeypair.privateKey);
    const requestBody = {
      header: {
        operation: 'create',
        kid: '#primary',
        alg: 'ES256K',
      },
      payload: encodedPayload,
      signature,
    };
    const encodedOperation = element.func.requestBodyToEncodedOperation({
      ...requestBody,
    });

    expect(encodedOperation).toMatchSnapshot();

    const anchorFileHash = await element.func.operationsToAnchorFile({
      operations: [encodedOperation],
      storage,
    });
    expect(anchorFileHash).toBe('QmXtnsawdNkTMBUWEBR1kTTP1gefcQiWG6aJQVq9pgrqNy');
    const anchorFile = await storage.read('QmXtnsawdNkTMBUWEBR1kTTP1gefcQiWG6aJQVq9pgrqNy');
    expect(anchorFile.didUniqueSuffixes[0]).toBe('MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI');
  });

  it('update', async () => {
    const encodedPayload = element.func.encodeJson({
      didUniqueSuffix: 'MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI',
      previousOperationHash: 'MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI',
      patch: [
        {
          op: 'replace',
          path: '/publicKey/2',
          value: {
            id: '#secondary',
            type: 'Secp256k1VerificationKey2018',
            publicKeyHex: secondaryKeypair.publicKey,
          },
        },
      ],
    });
    const signature = element.func.signEncodedPayload(encodedPayload, primaryKeypair.privateKey);
    const requestBody = {
      header: {
        operation: 'update',
        kid: '#primary',
        alg: 'ES256K',
      },
      payload: encodedPayload,
      signature,
    };
    const encodedOperation = element.func.requestBodyToEncodedOperation({
      ...requestBody,
    });
    expect(encodedOperation).toMatchSnapshot();

    const anchorFileHash = await element.func.operationsToAnchorFile({
      operations: [encodedOperation],
      storage,
    });
    expect(anchorFileHash).toBe('QmcLhmvUqEinBgdjm3bbLUXjnVuR9uhZPFVztpH7UJZSZQ');
    const anchorFile = await storage.read('QmcLhmvUqEinBgdjm3bbLUXjnVuR9uhZPFVztpH7UJZSZQ');
    expect(anchorFile.didUniqueSuffixes[0]).toBe('MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI');
  });

  it('recover', async () => {
    const encodedPayload = element.func.encodeJson({
      didUniqueSuffix: 'MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI',
      previousOperationHash: 'MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI',
      patch: [
        // first op should update recovery key.
        {
          op: 'replace',
          path: '/publicKey/1',
          value: {
            id: '#recovery',
            type: 'Secp256k1VerificationKey2018',
            publicKeyHex: recoveryKeypair2.publicKey,
          },
        },
        {
          op: 'replace',
          path: '/publicKey/0',
          value: {
            id: '#primary',
            type: 'Secp256k1VerificationKey2018',
            publicKeyHex: primaryKeypair2.publicKey,
          },
        },
      ],
    });
    const signature = element.func.signEncodedPayload(encodedPayload, primaryKeypair.privateKey);
    const requestBody = {
      header: {
        operation: 'recover',
        kid: '#recovery',
        alg: 'ES256K',
      },
      payload: encodedPayload,
      signature,
    };
    const encodedOperation = element.func.requestBodyToEncodedOperation({
      ...requestBody,
    });
    expect(encodedOperation).toMatchSnapshot();

    const anchorFileHash = await element.func.operationsToAnchorFile({
      operations: [encodedOperation],
      storage,
    });
    expect(anchorFileHash).toBe('QmUzCXShDkfaoYYFLSubjw3oKAit8LJoYx1S5YvcghvvNr');
    const anchorFile = await storage.read('QmUzCXShDkfaoYYFLSubjw3oKAit8LJoYx1S5YvcghvvNr');
    expect(anchorFile.didUniqueSuffixes[0]).toBe('MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI');
  });

  it('delete', async () => {
    const encodedPayload = element.func.encodeJson({
      didUniqueSuffix: 'MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI',
    });
    const signature = element.func.signEncodedPayload(encodedPayload, primaryKeypair.privateKey);
    const requestBody = {
      header: {
        operation: 'delete',
        kid: '#primary',
        alg: 'ES256K',
      },
      payload: encodedPayload,
      signature,
    };
    const encodedOperation = element.func.requestBodyToEncodedOperation({
      ...requestBody,
    });
    expect(encodedOperation).toMatchSnapshot();

    const anchorFileHash = await element.func.operationsToAnchorFile({
      operations: [encodedOperation],
      storage,
    });
    expect(anchorFileHash).toBe('QmaDD5M6uczyWJD3ua5RpAttMPirEef7APLhrBexHqrEGc');
    const anchorFile = await storage.read('QmaDD5M6uczyWJD3ua5RpAttMPirEef7APLhrBexHqrEGc');
    expect(anchorFile.didUniqueSuffixes[0]).toBe('MRO_nAwc19U1pusMn5PXd_5iY6ATvCyeuFU-bO0XUkI');
  });
});
