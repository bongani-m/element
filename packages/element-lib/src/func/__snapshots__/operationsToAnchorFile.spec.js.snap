// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operationsToAnchorFile create 1`] = `"eyJoZWFkZXIiOnsib3BlcmF0aW9uIjoiY3JlYXRlIiwia2lkIjoiI3ByaW1hcnkiLCJhbGciOiJFUzI1NksifSwicGF5bG9hZCI6ImV5SkFZMjl1ZEdWNGRDSTZJbWgwZEhCek9pOHZkek5wWkM1dmNtY3ZaR2xrTDNZeElpd2ljSFZpYkdsalMyVjVJanBiZXlKcFpDSTZJaU53Y21sdFlYSjVJaXdpZEhsd1pTSTZJbE5sWTNBeU5UWnJNVlpsY21sbWFXTmhkR2x2Ymt0bGVUSXdNVGdpTENKd2RXSnNhV05MWlhsSVpYZ2lPaUl3TWpnMlkyUm1aV0pqTkRNMU1UbGpOalJpWVdZd09EQXhOMkk1WmpFMFlUWTJPRGN4WldVNU1EYzNOekJsTm1Kak9HWmpaRFl5TWpneVpXSmpOekprWXpBaWZTeDdJbWxrSWpvaUkzSmxZMjkyWlhKNUlpd2lkSGx3WlNJNklsTmxZM0F5TlRack1WWmxjbWxtYVdOaGRHbHZia3RsZVRJd01UZ2lMQ0p3ZFdKc2FXTkxaWGxJWlhnaU9pSXdNak0wT0RNMU16VXlaV1EyT1RJMFltVTJZV1ExWlRKallXVmtNR0V6Tmpaak9USTVPRGRrWkdGbFpEVXlZVGMyWm1FNU1UWXdNamcwTm1Gak9HWmtZVE1pZlYxOSIsInNpZ25hdHVyZSI6IlpSeVZhOFVUSGF1bmxVSXQ2dFFIUVlxMVhDZVc2UXh3eWE3T2VvTEhJZlllN3RmZWR0dGFwTTJvNVRqa3cwOVlyaUdmQVo5end1ZnpXX1RLNHgxMlpBIn0"`;

exports[`operationsToAnchorFile delete 1`] = `"eyJoZWFkZXIiOnsib3BlcmF0aW9uIjoiZGVsZXRlIiwia2lkIjoiI3ByaW1hcnkiLCJhbGciOiJFUzI1NksifSwicGF5bG9hZCI6ImV5SmthV1JWYm1seGRXVlRkV1ptYVhnaU9pSk5VazlmYmtGM1l6RTVWVEZ3ZFhOTmJqVlFXR1JmTldsWk5rRlVka041WlhWR1ZTMWlUekJZVld0SkluMCIsInNpZ25hdHVyZSI6InRiRWRuVmdvOXRCdGN6cWViLV9mQzZfZ2tvMkVSOG1ZWlU0amRmNk8xOGsxSDJidlFuSFdrb0dKeXpNZnVuNS1RLU8xd1JJS20wcDhncWZDQkJOQ2pRIn0"`;

exports[`operationsToAnchorFile recover 1`] = `"eyJoZWFkZXIiOnsib3BlcmF0aW9uIjoicmVjb3ZlciIsImtpZCI6IiNyZWNvdmVyeSIsImFsZyI6IkVTMjU2SyJ9LCJwYXlsb2FkIjoiZXlKa2FXUlZibWx4ZFdWVGRXWm1hWGdpT2lKTlVrOWZia0YzWXpFNVZURndkWE5OYmpWUVdHUmZOV2xaTmtGVWRrTjVaWFZHVlMxaVR6QllWV3RKSWl3aWNISmxkbWx2ZFhOUGNHVnlZWFJwYjI1SVlYTm9Jam9pVFZKUFgyNUJkMk14T1ZVeGNIVnpUVzQxVUZoa1h6VnBXVFpCVkhaRGVXVjFSbFV0WWs4d1dGVnJTU0lzSW5CaGRHTm9JanBiZXlKdmNDSTZJbkpsY0d4aFkyVWlMQ0p3WVhSb0lqb2lMM0IxWW14cFkwdGxlUzh4SWl3aWRtRnNkV1VpT25zaWFXUWlPaUlqY21WamIzWmxjbmtpTENKMGVYQmxJam9pVTJWamNESTFObXN4Vm1WeWFXWnBZMkYwYVc5dVMyVjVNakF4T0NJc0luQjFZbXhwWTB0bGVVaGxlQ0k2SWpBelpUUmxZalF5WVdKa01tTTVPR0U1TURZellUbGhaakV6WW1GaE5tUmtNV1UxWm1RNFpUSXlNMll4TmpSbU5XUXpOakkwTXpsaE0yRTJOekF3WlRJMk9DSjlmU3g3SW05d0lqb2ljbVZ3YkdGalpTSXNJbkJoZEdnaU9pSXZjSFZpYkdsalMyVjVMekFpTENKMllXeDFaU0k2ZXlKcFpDSTZJaU53Y21sdFlYSjVJaXdpZEhsd1pTSTZJbE5sWTNBeU5UWnJNVlpsY21sbWFXTmhkR2x2Ymt0bGVUSXdNVGdpTENKd2RXSnNhV05MWlhsSVpYZ2lPaUl3TWpnek1UWmhNakJrWkRZNVl6WXdObVJtTm1RMVpURXpNREV6WkRFMU1XUTNOREUwT0RFMFpqUXhOMk14TXpBM09XSmxaRFpqWkRCbFpUUmtPR05oTTJRaWZYMWRmUSIsInNpZ25hdHVyZSI6IlpiOE9mY0RteEpLdUdzdFgxSkh1dUJMU09yWmFsWTQ2VnRjXzRaSmx6Wmw3WG8yS291ektwTUFvWnBqRVhmOWZiTGVZTXBiWVFZVW1nTVVhbC11cGJBIn0"`;

exports[`operationsToAnchorFile update 1`] = `"eyJoZWFkZXIiOnsib3BlcmF0aW9uIjoidXBkYXRlIiwia2lkIjoiI3ByaW1hcnkiLCJhbGciOiJFUzI1NksifSwicGF5bG9hZCI6ImV5SmthV1JWYm1seGRXVlRkV1ptYVhnaU9pSk5VazlmYmtGM1l6RTVWVEZ3ZFhOTmJqVlFXR1JmTldsWk5rRlVka041WlhWR1ZTMWlUekJZVld0Sklpd2ljSEpsZG1sdmRYTlBjR1Z5WVhScGIyNUlZWE5vSWpvaVRWSlBYMjVCZDJNeE9WVXhjSFZ6VFc0MVVGaGtYelZwV1RaQlZIWkRlV1YxUmxVdFlrOHdXRlZyU1NJc0luQmhkR05vSWpwYmV5SnZjQ0k2SW5KbGNHeGhZMlVpTENKd1lYUm9Jam9pTDNCMVlteHBZMHRsZVM4eUlpd2lkbUZzZFdVaU9uc2lhV1FpT2lJamMyVmpiMjVrWVhKNUlpd2lkSGx3WlNJNklsTmxZM0F5TlRack1WWmxjbWxtYVdOaGRHbHZia3RsZVRJd01UZ2lMQ0p3ZFdKc2FXTkxaWGxJWlhnaU9pSXdNelV6WldObE5qQTJZekZpWlRsbE1qZzBNVFpqWkdJMFlUVXhaamM0WkdJeE0yVm1ZV0l4WWpVMU9ETTNZekZtWVRVNE5HUTBZek0yWWpBeFpERXdaREFpZlgxZGZRIiwic2lnbmF0dXJlIjoiZEUxSU5ZOVpaWEVKdVhzbF9IdDJERDNjeTNkLVVRaGJJcm5ibkRqWHVpa2E1SGVXVUxHcGlMU3NWaDc1T1lQMDZLRG9HWG5DLWpweXdzSW0waFlVOHcifQ"`;
